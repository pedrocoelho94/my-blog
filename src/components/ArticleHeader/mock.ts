import { data } from 'api/mockPosts.json'

const { id, title, excerpt, cover, author, categories, createdAt } =
  data.posts[0]

export default {
  id,
  title,
  excerpt,
  cover,
  author,
  categories,
  createdAt
}
