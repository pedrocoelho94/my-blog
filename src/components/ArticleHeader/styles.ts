import styled, { css } from 'styled-components'
import * as HeadingStyles from 'components/Heading/styles'

export const Wrapper = styled.header`
  ${({ theme }) => css`
    padding-bottom: ${theme.spacings.xlarge};
    margin-bottom: ${theme.spacings.xlarge};
    border-bottom: 0.1rem solid ${theme.colors.mediumGray};

    ${HeadingStyles.Title} {
      margin: 0;
      margin-bottom: ${theme.spacings.medium};
      font-size: ${theme.font.sizes.large};
    }
  `}
`

export const Excerpt = styled.p`
  ${({ theme }) => css`
    margin: ${theme.spacings.medium} 0;
    font-size: calc(${theme.font.sizes.medium} * 0.9);
  `}
`

export const Cover = styled.img`
  ${({ theme }) => css`
    max-width: 100%;
    display: block;
    margin-bottom: ${theme.spacings.medium};
  `}
`
