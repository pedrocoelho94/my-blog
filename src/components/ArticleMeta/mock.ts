export default {
  createdAt: '2021-03-02T15:16:09.246+00:00',
  author: {
    id: '123456789',
    slug: 'pedro-coelho',
    displayName: 'Pedro Coelho'
  },
  categories: [
    {
      id: '78724ec23f',
      displayName: 'Tech',
      slug: 'tech'
    },
    {
      id: '762vg214g34',
      displayName: 'JS',
      slug: 'javascript'
    }
  ]
}
