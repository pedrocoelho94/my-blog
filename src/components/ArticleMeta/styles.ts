import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.darkerGray};
    font-style: italic;

    .categories span:not(:last-child):after {
      content: ', ';
    }

    a {
      color: ${theme.colors.secondary};
      text-decoration: none;
      transition: all 0.2s ease-in-out;

      &:hover {
        filter: brightness(0.5);
      }
    }
  `}
`
