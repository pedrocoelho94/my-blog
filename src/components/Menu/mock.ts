export default {
  links: [
    {
      id: 'link-1',
      link: '/',
      newTab: false,
      text: 'Home'
    },
    {
      id: 'link-2',
      link: '/news',
      newTab: false,
      text: 'Notícias'
    },
    {
      id: 'link-3',
      link: '/reviews',
      newTab: false,
      text: 'Críticas'
    },
    {
      id: 'link-4',
      link: '/entrevistas',
      newTab: false,
      text: 'Entrevistas'
    },
    {
      id: 'link-5',
      link: '/about',
      newTab: false,
      text: 'Sobre'
    }
  ],
  blogName: 'Claquete 42',
  logo: '/assets/images/claquete.png'
}
