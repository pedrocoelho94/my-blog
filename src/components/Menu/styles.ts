import styled, { css } from 'styled-components'
import media from 'styled-media-query'

type MenuBehavior = {
  isMenuVisible: boolean
}

export const Wrapper = styled.div<MenuBehavior>`
  ${({ isMenuVisible }) => css`
    z-index: 9;
    transition: background 0.2s ease-in-out;

    ${isMenuVisible &&
    css`
      position: fixed;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      background: rgba(0, 0, 0, 0.5);
    `};
  `}
`

export const Content = styled.main<MenuBehavior>`
  ${({ theme, isMenuVisible }) => css`
    background: ${theme.colors.primary};
    padding: ${theme.spacings.large};
    display: flex;
    position: fixed;
    width: 100%;
    max-width: 26rem;
    height: 100vh;
    top: 0;
    transition: all 0.3s ease-in-out;

    left: ${isMenuVisible ? '0' : '-24rem'};
    overflow-y: ${isMenuVisible ? 'auto' : 'hidden'};

    ${media.lessThan('medium')`
      left: ${isMenuVisible ? '0' : '-26rem'};
    `}
  `}
`

export const Nav = styled.nav`
  ${({ theme }) => css`
    margin: ${theme.spacings.xxlarge} auto;
    width: 100%;
  `}
`

export const OpenClose = styled.a<MenuBehavior>`
  ${({ theme, isMenuVisible }) => css`
    position: fixed;
    top: ${theme.spacings.medium};
    background: ${theme.colors.primary};
    z-index: 2;
    width: 3rem;
    height: 3rem;
    transition: all 0.3s ease-in-out;
    left: 0.5rem;
    top: 2rem;
    color: ${theme.colors.white};

    ${!isMenuVisible &&
    css`
      ${media.lessThan('medium')`
        left: 0rem;
      `}
    `}

    ${isMenuVisible &&
    css`
      left: 21rem;
      color: ${theme.colors.secondary};
    `};
  `}
`
