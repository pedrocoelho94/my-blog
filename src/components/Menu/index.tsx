import { useRouter } from 'next/router'
import Link from 'next/link'

import * as S from './styles'
import MenuLink from 'components/MenuLink'
import { useEffect, useRef, useState } from 'react'

import { Search } from '@styled-icons/material-outlined'

import { signOut, useSession } from 'next-auth/react'

export type MenuPropsLinks = {
  id: string
  link: string
  newTab?: boolean
  text: string
}

export type MenuProps = {
  links: MenuPropsLinks[]
}

const Menu = ({ links = [] }: MenuProps) => {
  const { data: session } = useSession()

  const router = useRouter()
  const [isAtTop, setIsAtTop] = useState(true)
  const [searchValue, setSearchValue] = useState('')
  const [toggle, setToggleNav] = useState(false)
  const searchBarRef = useRef<HTMLInputElement>(null)

  // Diminui o tamanho do menu depois de descer 300px
  const changeHeightBar = () => {
    if (window.pageYOffset > 300) {
      setIsAtTop(false)
    } else {
      setIsAtTop(true)
    }
  }

  useEffect(() => {
    window.addEventListener('scroll', changeHeightBar)

    return () => {
      window.removeEventListener('scroll', changeHeightBar)
    }
  }, [])

  useEffect(() => {
    setSearchValue('')
    setToggleNav(false)
    searchBarRef.current?.blur()
  }, [router?.query.slug])

  const handleSearch = (e: React.FormEvent) => {
    e.preventDefault()

    if (router.query.slug === searchValue) {
      setSearchValue('')
      setToggleNav(false)
    }
    router.push(`/search/${encodeURI(searchValue)}`)
  }

  const handleClick = () => {
    setToggleNav(false)
  }

  const handleClickLogin = (event: React.MouseEvent) => {
    event.preventDefault()
    signOut({ redirect: false })
  }

  return (
    <S.Wrapper isAtTop={isAtTop} aria-label="navbar">
      <S.Content>
        <S.LogoContainer>
          <Link href="/" passHref>
            <a>
              <S.Logo src="/assets/images/claquete.png" />
            </a>
          </Link>
        </S.LogoContainer>

        <S.Menu role="navigation">
          {links.map((link) => (
            <S.Item key={`web-${link.id}`}>
              <MenuLink newTab={link.newTab} link={link.link}>
                {link.text}
              </MenuLink>
            </S.Item>
          ))}

          {session ? (
            <S.Dropdown>
              <S.Dropbtn>
                <MenuLink link="#">{session.user.name}</MenuLink>
              </S.Dropbtn>
              <S.DropdownContent>
                <MenuLink link="/profile">Perfil</MenuLink>
                <MenuLink link="/profile/posts">Meus Posts</MenuLink>
                <MenuLink link="/profile/create-post">Novo Post</MenuLink>
                <div onClick={handleClickLogin}>
                  <MenuLink link="#">Sair</MenuLink>
                </div>
              </S.DropdownContent>
            </S.Dropdown>
          ) : (
            <MenuLink link="/login">Login</MenuLink>
          )}

          <S.SearchBox onSubmit={handleSearch} role="search">
            <S.SearchLabel htmlFor="search">Buscar</S.SearchLabel>
            <S.SearchInput
              id="search"
              placeholder="Buscar..."
              value={searchValue}
              onChange={(e) => setSearchValue(e.target.value)}
              aria-label="search box"
              ref={searchBarRef}
            />
            <S.SearchButton type="submit">
              <Search size={22} />
            </S.SearchButton>
          </S.SearchBox>
        </S.Menu>

        <S.NavIcon
          onClick={() => setToggleNav(!toggle)}
          aria-hidden={!toggle}
          aria-label="open or close menu"
          title={toggle ? 'Fechar menu' : 'Abrir menu'}
        >
          <S.Line open={toggle} />
          <S.Line open={toggle} />
          <S.Line open={toggle} />
        </S.NavIcon>
      </S.Content>

      <S.Overlay open={toggle} aria-label="menu mobile" aria-hidden={!toggle}>
        <S.OverlayMenu open={toggle}>
          {links.map((link) => (
            <S.Item key={`mobile-${link.id}`}>
              <div onClick={handleClick} aria-label="closeMenuDiv">
                <MenuLink newTab={link.newTab} link={link.link}>
                  {link.text}
                </MenuLink>
              </div>
            </S.Item>
          ))}

          <S.SearchBoxMobile onSubmit={handleSearch} open={toggle}>
            <S.SearchLabel htmlFor="searchMobile">Buscar</S.SearchLabel>
            <S.SearchInput
              id="searchMobile"
              placeholder="Buscar"
              value={searchValue}
              onChange={(e) => setSearchValue(e.target.value)}
              aria-label="search box"
            />
            <S.SearchButton type="submit">
              <Search size={24} />
            </S.SearchButton>
          </S.SearchBoxMobile>
        </S.OverlayMenu>
      </S.Overlay>
    </S.Wrapper>
  )
}

export default Menu
