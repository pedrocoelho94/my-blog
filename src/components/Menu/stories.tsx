import { Story, Meta } from '@storybook/react'
import Menu, { MenuProps } from '.'
import mock from './mock'

export default {
  title: 'Menu',
  component: Menu,
  args: mock
} as Meta

export const Default: Story<MenuProps> = (args) => (
  <div>
    <Menu {...args} />
  </div>
)
