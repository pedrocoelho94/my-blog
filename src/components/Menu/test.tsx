import { screen, fireEvent } from '@testing-library/react'
import { SessionProvider } from 'next-auth/react'
import { renderWithTheme } from 'styles/renderWithTheme'

import Menu, { MenuProps } from '.'
import mock from './mock'
const props: MenuProps = mock

const sessionProps = {
  expires: '1',
  user: { email: 'a', name: 'Delta', image: 'c' }
}

describe('<Menu />', () => {
  it('should render full menu', () => {
    renderWithTheme(
      <SessionProvider session={sessionProps}>
        <Menu {...props} />
      </SessionProvider>
    )

    expect(screen.getByRole('navigation')).toBeInTheDocument()
    expect(screen.getByLabelText(/open or close menu/i)).toHaveStyleRule(
      'display',
      'none',
      { media: '(min-width: 768px)' }
    )
    expect(screen.getAllByLabelText(/search box/i)).toHaveLength(2)

    expect(screen.getByLabelText('navbar')).toHaveStyle({
      height: '8rem'
    })
  })

  it('should open/close menu on button click', () => {
    renderWithTheme(
      <SessionProvider session={sessionProps}>
        <Menu {...props} />
      </SessionProvider>
    )

    // muda o viewport do jest
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: 500
    })
    window.dispatchEvent(new Event('resize'))
    expect(window.innerWidth).toBe(500)

    const iconButton = screen.getByLabelText(/open or close menu/i)
    const menuMobile = screen.getByLabelText(/menu mobile/i)

    expect(iconButton).toHaveStyle({
      display: 'list-item'
    })

    expect(menuMobile.getAttribute('aria-hidden')).toBe('true')

    // Menu abre após o clique no icone do menu
    fireEvent.click(iconButton)
    expect(menuMobile.getAttribute('aria-hidden')).toBe('false')

    // Menu fecha após o clique no icone do menu
    fireEvent.click(iconButton)
    expect(menuMobile.getAttribute('aria-hidden')).toBe('true')
  })
})
