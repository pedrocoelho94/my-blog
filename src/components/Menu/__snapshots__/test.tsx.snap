// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should match snapshot 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  z-index: 9;
  -webkit-transition: background 0.2s ease-in-out;
  transition: background 0.2s ease-in-out;
}

.c3 {
  background: #000000;
  padding: 3.2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  width: 100%;
  max-width: 26rem;
  height: 100vh;
  top: 0;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  left: -24rem;
  overflow-y: hidden;
}

.c4 {
  margin: 4.8rem auto;
  width: 100%;
}

.c1 {
  position: fixed;
  top: 2.4rem;
  background: #000000;
  z-index: 2;
  width: 3rem;
  height: 3rem;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  left: 0.5rem;
  top: 2rem;
  color: #FFFFFF;
}

.c5 {
  display: block;
  color: #FFFFFF;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin-bottom: 1.6rem;
  font-size: 1.8rem;
  border-right: 0.5rem solid #000000;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
}

.c5:hover {
  border-color: #dc143c;
  color: #dc143c;
}

@media (max-width:768px) {
  .c3 {
    left: -26rem;
  }
}

@media (max-width:768px) {
  .c1 {
    left: 0rem;
  }
}

<div>
  <div
    class="c0"
  >
    <a
      aria-label="Open or close menu"
      class="c1"
      href="#"
      title="Open menu"
    >
      <svg
        aria-hidden="true"
        aria-label="Open menu"
        class="c2"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </a>
    <main
      aria-hidden="true"
      class="c3"
    >
      <nav
        class="c4"
      >
        <a
          class="c5"
          href="/home/1"
          target="_self"
        >
          My jobs
        </a>
        <a
          class="c5"
          href="/home/2"
          target="_self"
        >
          About me
        </a>
        <a
          class="c5"
          href="/home/3"
          target="_self"
        >
          Services
        </a>
        <a
          class="c5"
          href="http://localhost/home/4"
          target="_self"
        >
          External Link
        </a>
        <a
          class="c5"
          href="/home/5"
          target="_self"
        >
          JavaScript
        </a>
        <a
          class="c5"
          href="/home/6"
          target="_self"
        >
          Python
        </a>
        <a
          class="c5"
          href="/home/7"
          target="_self"
        >
          Consectetur
        </a>
        <a
          class="c5"
          href="/home/8"
          target="_self"
        >
          Text test
        </a>
        <a
          class="c5"
          href="/home/9"
          target="_self"
        >
          Pac-man
        </a>
        <a
          class="c5"
          href="/home/10"
          target="_self"
        >
          Star Wars
        </a>
        <a
          class="c5"
          href="/home/11"
          target="_self"
        >
          A menu
        </a>
      </nav>
    </main>
  </div>
</div>
`
